### PR TITLE
Log redis failures during operations 

### LINKED ISSUE #2610: Failed writes to Redis are potentially ignored
The Redis adapter in the ui-server returns true or false [here](https://github.com/SwissDataScienceCenter/renku-ui/blob/master/server/src/storage/RedisStorage.ts#L97).

But in several places the result from the `save` method is not checked. And the code assumes that the write to Redis was successful. Or perhaps Redis raises an error if the write was not successful. If Redis raises an error on failed writes then it would be good to check the code and ensure that ignoring the return value from `storage.save` is correct or some checks should be added. Or if the true/false value being returned from the save method indicates a successful write to Redis then we should be always checking the output and raising an appropriate error.

Here is where this happens:
- https://github.com/SwissDataScienceCenter/renku-ui/blob/master/server/src/authentication/index.ts#L182
- https://github.com/SwissDataScienceCenter/renku-ui/blob/master/server/src/authentication/index.ts#L189
- https://github.com/SwissDataScienceCenter/renku-ui/blob/master/server/src/authentication/index.ts#L365 (indirectly but the return value is still ignored)
- https://github.com/SwissDataScienceCenter/renku-ui/blob/master/server/src/utils/middlewares/lastProjectsMiddleware.ts#L52

There are a few more cases.

### PR DESCRIPTION
**Pull request by ciyer**
Check return values from redis and log errors when appropriate.

Fix #2610

/deploy #persist renku=renku-ui/3.11-acceptance-tests

### CODE REVIEW

File: server/src/utils/middlewares/lastProjectsMiddleware.ts [hunk 1]
```diff
@@ -49,15 +52,22 @@ const lastProjectsMiddleware =
...
+          .then((value) => {
+            if (!value) {
+              const errorMessage = `Error saving project ${projectName} for user ${userId}`;
+              logger.error(errorMessage);
+              Sentry.captureMessage(errorMessage);
+            }
+          })
+          .catch((err) => {
+            Sentry.captureException(err);
+          });
```
Comment (lorenzo-cavazzi) [inline server/src/utils/middlewares/lastProjectsMiddleware.ts:70]: The code here makes sense to me, but when I try to simulate a storage failure just before navigating to a project from the UI (I manually delete all the Redis instances from Rancher) I get a `500 internal error` response; from the code, I assume it should log the error to Sentry and jump to the next step.

image

Perhaps something with `storage.save` is off?

**Review by lorenzo-cavazzi — CHANGES_REQUESTED** [review]
The code makes sense to me, but I spotted a strange case with the "lastVisistedProjects" middleware (more info inline).

After a few more trials, I wonder if the problem is in the `RedisStorage` class. I got the following error when simulating a failure; we are not catching exceptions from `ioredis`, perhaps we should wrap `this.redis.<someAction>` around a `try/catch`

```
[start:dev-debug] [ioredis] Unhandled error event: Error: All sentinels are unreachable. Retrying from scratch after 10ms. Last error: Connection is closed.
[start:dev-debug]     at SentinelConnector.<anonymous> (/home/lorenzo/development/sdsc/renku-ui/server/node_modules/ioredis/built/connectors/SentinelConnector/index.js:73:31)
[start:dev-debug]     at Generator.next (<anonymous>)
[start:dev-debug]     at /home/lorenzo/development/sdsc/renku-ui/server/node_modules/ioredis/built/connectors/SentinelConnector/index.js:8:71
[start:dev-debug]     at new Promise (<anonymous>)
[start:dev-debug]     at __awaiter (/home/lorenzo/development/sdsc/renku-ui/server/node_modules/ioredis/built/connectors/SentinelConnector/index.js:4:12)
[start:dev-debug]     at connectToNext (/home/lorenzo/development/sdsc/renku-ui/server/node_modules/ioredis/built/connectors/SentinelConnector/index.js:59:37)
[start:dev-debug]     at SentinelConnector.<anonymous> (/home/lorenzo/development/sdsc/renku-ui/server/node_modules/ioredis/built/connectors/SentinelConnector/index.js:125:24)
[start:dev-debug]     at Generator.throw (<anonymous>)
[start:dev-debug]     at rejected (/home/lorenzo/development/sdsc/renku-ui/server/node_modules/ioredis/built/connectors/SentinelConnector/index.js:6:65)
```

**Comment by leafty** [comment]
> The code makes sense to me, but I spotted a strange case with the "lastVisistedProjects" middleware (more info inline).
> 
> After a few more trials, I wonder if the problem is in the `RedisStorage` class. I got the following error when simulating a failure; we are not catching exceptions from `ioredis`, perhaps we should wrap `this.redis.<someAction>` around a `try/catch`
> 
> ```
> [start:dev-debug] [ioredis] Unhandled error event: Error: All sentinels are unreachable. Retrying from scratch after 10ms. Last error: Connection is closed.
> [start:dev-debug]     at SentinelConnector.<anonymous> (/home/lorenzo/development/sdsc/renku-ui/server/node_modules/ioredis/built/connectors/SentinelConnector/index.js:73:31)
> [start:dev-debug]     at Generator.next (<anonymous>)
> [start:dev-debug]     at /home/lorenzo/development/sdsc/renku-ui/server/node_modules/ioredis/built/connectors/SentinelConnector/index.js:8:71
> [start:dev-debug]     at new Promise (<anonymous>)
> [start:dev-debug]     at __awaiter (/home/lorenzo/development/sdsc/renku-ui/server/node_modules/ioredis/built/connectors/SentinelConnector/index.js:4:12)
> [start:dev-debug]     at connectToNext (/home/lorenzo/development/sdsc/renku-ui/server/node_modules/ioredis/built/connectors/SentinelConnector/index.js:59:37)
> [start:dev-debug]     at SentinelConnector.<anonymous> (/home/lorenzo/development/sdsc/renku-ui/server/node_modules/ioredis/built/connectors/SentinelConnector/index.js:125:24)
> [start:dev-debug]     at Generator.throw (<anonymous>)
> [start:dev-debug]     at rejected (/home/lorenzo/development/sdsc/renku-ui/server/node_modules/ioredis/built/connectors/SentinelConnector/index.js:6:65)
> ```

Just a non-blocking remark here: wrappers like the RedisStorage class are usually prone to create bugs and issues over time. I would rather use the redis client library directly and handle error states at the call site.

**Comment by lorenzo-cavazzi** [comment]
> wrappers like the RedisStorage class are usually prone to create bugs and issues over time.

I agree with this.
The main driver to have a wrapper was the idea to move away from Redis soon (delegate auth to Gateway, and use permanent storage for everything else). How confusing are responses from `ioresdis` was another point for that solution. Ofc that played out badly and in insight I would not go for this solution again.

At this point we will probably repurpose the UI server soon, I would say we can keep the current approach for the time being.

**Review by lorenzo-cavazzi — APPROVED** [review]
Lgtm! :rocket:

**Pull request merged**